sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala

```diff
@@ -61,6 +61,9 @@ case class SortOrder(child: Expression, direction: SortDirection)
   override def sql: String = child.sql + " " + direction.sql
 
   def isAscending: Boolean = direction == Ascending
+
+  def semanticEquals(other: SortOrder): Boolean =
```

Comment:
@cloud-fan : If you look at the old version of `EnsureRequirements` below at L253, it compared raw `SortOrder` objects which will use `equals()` generated for it. In scala, `equals()` for case classes is merely doing `equals()` over all its fields so that lead to `Expression`'s `equals()` being used instead of its `semanticEquals()`.

My fix here was to introduce a `semanticEquals` in `SortOrder` which compares the underlying `Expression` semantically.